File: 0054-spiral-matrix/0054-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let ans = [];
    let sr = 0;
    let er = matrix.length-1;
    let sc = 0;
    let ec = matrix[0].length-1;
    while(sr <= er && sc <= ec){
        //top
         for(let j = sc; j <= ec; j++){
            ans.push(matrix[sr][j]);
         } 
         sr++;
        //right
        for(let j = sr; j <= er; j++){
            ans.push(matrix[j][ec]);
         } 
         ec--;
        //down 
        if(sr <= er){
            for(let j = ec ; j >= sc; j--){
            ans.push(matrix[er][j]);
         }
          er--;
        }
        //left
        if(sc <= ec){
            for(let j = er; j >= sr; j--){
             ans.push(matrix[j][sc]);
         } 
        sc++;
    }
        }
         
   return ans;
};

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        List<Integer> l = new ArrayList<>();
        int n = matrix[0].length;
        int sr = 0;
        int sc = 0;
        int er = m-1;
        int ec = n-1;
        while(sr <= er && sc <= ec){
            //right
            for(int i = sc; i <= ec ; i++){
                l.add(matrix[sr][i]);
            }
            sr++;
            //down
            for(int i = sr; i <= er; i++){
                l.add(matrix[i][ec]);
            }
            ec--;

            //left
            if(er >= sr){
                for(int i = ec; i >= sc; i--){
                l.add(matrix[er][i]);
            }
            er--;

            }
            
            //up
            if(ec >= sc ){
                for(int i = er; i >= sr; i--){
                    l.add(matrix[i][sc]);
                }
                sc++;
            }
        }
        return l;
    }
}